velox/expression/Expr.cpp

```diff
@@ -1399,128 +1252,40 @@ void setPeeledArg(
-        if (!firstIndices) {
-          firstIndices = std::move(indices);
-        } else if (indices != firstIndices) {
-          // different fields use different dictionaries
-          peeled = false;
-          break;
-        }
-        if (!firstWrapper) {
-          firstWrapper = leaf;
-        }
-        setPeeledArg(leaf->valueVector(), i, numArgs, maybePeeled);
-      } else if (encoding == VectorEncoding::Simple::SEQUENCE) {
-        if (firstIndices) {
-          // having a mix of dictionary and sequence encoded fields
-          peeled = false;
-          break;
-        }
-        BufferPtr lengths = leaf->wrapInfo();
-        if (!firstLengths) {
-          firstLengths = std::move(lengths);
-        } else if (lengths != firstLengths) {
-          // different fields use different sequences
-          peeled = false;
-          break;
-        }
-        if (!firstWrapper) {
-          firstWrapper = leaf;
-        }
-        setPeeledArg(leaf->valueVector(), i, numArgs, maybePeeled);
-      } else {
-        // Non-peelable encoding.
-        peeled = false;
-        break;
-      }
-    }
-    if (peeled) {
-      ++numLevels;
-      inputValues_ = std::move(maybePeeled);
-    }
-  } while (peeled && numConstant != numArgs);
-  if (!numLevels) {
-    return false;
-  }
+  LocalDecodedVector localDecoded(context);
   LocalSelectivityVector newRowsHolder(context);
   ScopedContextSaver saver;
-  // We peel off the wrappers and make a new selection.
-  SelectivityVector* newRows;
-  LocalDecodedVector localDecoded(context);
-  if (numConstant == numArgs) {
-    // All the fields are constant across the rows of interest.
-    newRows = singleRow(newRowsHolder, rows.begin());
-
-    context.saveAndReset(saver, applyRows);
-    context.setConstantWrap(rows.begin());
-  } else {
-    auto decoded = localDecoded.get();
-    decoded->makeIndices(*firstWrapper, rows, numLevels);
-    newRows = translateToInnerRows(applyRows, *decoded, newRowsHolder);
-    context.saveAndReset(saver, applyRows);
-    setDictionaryWrapping(*decoded, rows, *firstWrapper, context);
-
-    // 'newRows' comes from the set of row numbers in the base vector. These
-    // numbers may be larger than rows.end(). Hence, we need to resize constant
-    // inputs.
-    if (newRows->end() > rows.end() && numConstant) {
-      for (int i = 0; i < constantArgs.size(); ++i) {
-        if (!constantArgs.empty() && constantArgs[i]) {
-          inputValues_[i] =
-              BaseVector::wrapInConstant(newRows->end(), 0, inputValues_[i]);
-        }
-      }
-    }
+  // Attempt peeling.
+  auto peeledEncoding = std::make_shared<PeeledEncoding>();
+  auto peeledVectors = peeledEncoding->getPeeledVectors(
+      inputValues_,
+      applyRows,
+      localDecoded,
+      vectorFunction_->isDefaultNullBehavior());
+  if (isFlat(peeledEncoding->getWrapEncoding())) {
+    return false;
   }
+  inputValues_ = std::move(peeledVectors);
+  peeledVectors.clear();
```

Comment:
Once you move peeledVectors, is calling peeledVectors even for clearing it right ? will it be valid object (might it segfault ?)